[ CDM-243 ] [ CDM-245 ] Orcid Provider MFA Support & Token response `mfa` key

dauglyon opened the pull request:
Adds MFA boolean to RemoteIdentity and StoredToken classes, which get stored in Mongo. Adds JWT parsing to extract the AMR claim from the Orcid login token during login. Modifies the token to include MFA status. 

## Summary of Changes

### Core Implementation
  - Added `mfa` field to StoredToken (stored as mfa in MongoDB)
  - New `MfaStatus` enum with `@JsonValue` for consistent ID serialization
  - Extended RemoteIdentityDetails to capture MFA status from identity providers
  - Updated APIToken to expose MFA status via /api/V2/token endpoint

### ORCID Provider Enhancement
  - Changed scope from /authenticate to openid to receive JWT ID tokens
  - Parse Authentication Method Reference (AMR) claim to detect MFA usage
  - Set mfa=USED when AMR contains "mfa"
  - Supports both ORCID member (OpenID) and non-member (OAuth2) applications

### Authentication Flow
  - OAuth login: MFA status determined from identity provider and stored on LOGIN token
  - Password login: MFA status set to UNKNOWN (not applicable)
  - Agent/Dev/Service tokens: MFA status set to UNKNOWN (no authentication context)

### API Response

  The /api/V2/token endpoint now returns `mfa` with tri-state semantics:
  - `USED`: User authenticated with MFA during token creation (e.g. Orcid with MFA)
  - `NOT_USED`: User explicitly chose not to use MFA when available (e.g. Orcid without MFA)
  - `UNKNOWN`: MFA status unknown or not applicable to authentication method (e.g. Google (mfa inspection not supported), a non-member Orcid API client account (openid token not supported), or a dev token)

### Database Schema
  - Tokens collection: Added mfa field

dauglyon committed:
Add openid role and mfa parsing to Orcid provider; support MFA status; add MFA tests

dauglyon commented on src/main/java/us/kbase/auth2/lib/identity/RemoteIdentityDetails.java:
```diff
@@ -57,13 +75,22 @@ public String getFullname() {
 		int result = 1;
 		result = prime * result + ((email == null) ? 0 : email.hashCode());
 		result = prime * result + ((fullname == null) ? 0 : fullname.hashCode());
+		result = prime * result + ((mfaAuthenticated == null) ? 0 : mfaAuthenticated.hashCode());
```
Not sure this is right, as it is a boolean.

dauglyon replied:
(though it seems to be working)

dauglyon commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -228,6 +233,61 @@ private OrcIDAccessTokenResponse(
+					// Decode the payload (second part) - URL-safe base64
+					final String payload = new String(Base64.getUrlDecoder().decode(parts[1]));
+					
+					// Parse JSON payload to extract claims
```
Not sure this is the correct way to JSON in this codebase.

dauglyon commented on src/main/java/us/kbase/auth2/service/api/MfaStatus.java:
```diff
@@ -0,0 +1,89 @@
+	 */
+	@GET
+	@Produces(MediaType.APPLICATION_JSON)
+	public Map<String, Object> getMfaStatus(
```
Needs some changes so it doesn't inline the Orcid provider

dauglyon commented:
per @MrCreosote: don't add new endpoint for checking MFA, add new key to the token endpoint

dauglyon committed:
move MFA status to token endpoint; remove mfaStatus endpoint; clean up code

dauglyon committed:
fix build errors

dauglyon committed:
fix test failures from MFA implementation

Update tests to handle MFA field additions and ORCID OpenID Connect changes:
- Fix hash codes in RemoteIdentityTest for new MFA field
- Update ORCID provider tests for openid scope instead of /authenticate
- Add mfaAuthenticated field to API response expectations
- Fix non-ORCID provider test to use null MFA status
- Update identity ordering in user endpoint tests

dauglyon committed:
Improve test coverage; improve error handling

dauglyon committed:
fix exception errors

dauglyon committed:
Update ORCID provider tests to use MockServer

dauglyon committed:
remove duplicate tests

dauglyon committed:
fix TokenEndpointTest.getToken

dauglyon committed:
fix TokenEndpointTest.getToken

dauglyon committed:
Store MFA authentication status directly on tokens

  Move MFA status from being computed from user identities to being stored
  as a boolean field on tokens themselves. This provides per-token MFA
  tracking and eliminates the need for complex identity lookups.

dauglyon committed:
update tests

dauglyon commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -228,6 +233,58 @@ private OrcIDAccessTokenResponse(
+					return null;
+				} catch (IOException e) {
+					// JSON parsing failed - malformed payload
+					return null;
```
Could optionally throw an error in these cases, but we would need to handle the error during the login flow. Perhaps we just log this case and return a null MFA status? If the JWT is correct, we shouldn't see these. Thoughts?

dauglyon committed:
fix tests

dauglyon committed:
slightly better documentation

dauglyon committed:
update token field name

dauglyon committed:
consolidate test logic

dauglyon commented on src/test/java/us/kbase/test/auth2/service/api/UserEndpointTest.java:
Seems to be some stochasticity in these tests, as I had to change the order. I haven't identified a change I made that would cause this directly. Seen in src/test/resources/us/kbase/test/auth2/service/ui/MeTest_getMeMaximalInput.testdata, src/test/java/us/kbase/test/auth2/service/ui/MeTest.java, src/test/java/us/kbase/test/auth2/service/api/UserEndpointTest.java

dauglyon replied:
Regenerated hashcodes and this went away

MrCreosote commented on RELEASE_NOTES.md:
```diff
@@ -1,5 +1,11 @@
 # Authentication Service MKII release notes
 
+## 0.8.0
```
Since we're in a 0 major version, I use the minor version to denote backwards incompatible changes and patch for everything else. Based on that this should be 0.7.3

MrCreosote replied:
And actually, I realized we haven't released 0.7.2 yet, so a version bump isn't needed

MrCreosote replied:
This comment is still open

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
```diff
@@ -905,7 +911,9 @@ public NewToken createToken(
 				.withContext(tokenCtx)
-				.withTokenName(tokenName).build(),
+				.withTokenName(tokenName)
+				.withMfaAuthenticated(null) // Agent/Dev/Serv tokens don't have MFA status
```
Do we need the distinction between null and false for MFA?

dauglyon replied:
IMO, yes, it helps us know how to instruct a user to enable MFA (false, they need to enable it with their provider; null, the used login provider is not supported) 

We could represent this as three string instead, or an enum perhaps. But there are indeed three distinct states.

MrCreosote replied:
Ok, I would go with an Enum in that case, easier to read, no nulls.

MrCreosote replied:
The new Enum is all caps, whereas all the other enums that are exposed are CamelCase (I think, been a while since I touched this codebase):

https://github.com/kbase/auth2/blob/47f6a0160a6c3bec09f49a6623b3decb3d647a8d/src/main/java/us/kbase/auth2/lib/Role.java#L14
https://github.com/kbase/auth2/blob/47f6a0160a6c3bec09f49a6623b3decb3d647a8d/src/main/java/us/kbase/auth2/lib/token/TokenType.java#L11

Can we make the new enum the same?

MrCreosote replied:
I think this is similar to https://github.com/kbase/auth2/pull/471#discussion_r2505769481

dauglyon replied:
Enum updated with casing to match existing enums

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
```diff
@@ -745,13 +745,19 @@ public void forceResetAllPasswords(final IncomingToken token)
+	}
+	
+	private NewToken login(final UserName userName, final TokenCreationContext tokenCtx, 
+			final Boolean mfaAuthenticated) throws AuthStorageException {
```
mfaAuthenticated = multifactor authentication authenticated. Seems redundant

MrCreosote replied:
This pattern is copied all through the codebase

dauglyon replied:
Fair, will change this variable name

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
```diff
@@ -2043,6 +2051,7 @@ public NewToken testModeCreateToken(
 		final NewToken nt = new NewToken(StoredToken.getBuilder(tokenType, id, userName)
 				.withLifeTime(clock.instant(), TEST_MODE_DATA_LIFETIME_MS)
 				.withNullableTokenName(tokenName)
+				.withMfaAuthenticated(null) // Test mode tokens don't have MFA status
```
If this is the default state the builder should just set that state on initialization so the client code doesn't need to call this

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
```diff
@@ -2339,7 +2348,9 @@ public NewToken login(
 			}
 		}
-		return login(u.get().getUserName(), tokenCtx);
+		final Boolean mfaStatus = ri.get().getDetails() != null ? 
```
The RI details can't be null

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/identity/RemoteIdentityDetails.java:
```diff
@@ -57,13 +75,22 @@ public String getFullname() {
 		int result = 1;
 		result = prime * result + ((email == null) ? 0 : email.hashCode());
 		result = prime * result + ((fullname == null) ? 0 : fullname.hashCode());
+		result = prime * result + ((mfaAuthenticated == null) ? 0 : mfaAuthenticated.hashCode());
```
Boolean example: https://github.com/kbase/auth2/blob/develop/src/main/java/us/kbase/auth2/lib/user/LocalUser.java#L66

I just have eclipse autogenerate this stuff

MrCreosote replied:
Although that's for an unboxed boolean. Not sure I have an example for a boxed boolean, I avoid using those

MrCreosote replied:
This is moot since it's now an enum

:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/identity/RemoteIdentityDetails.java:
```diff
@@ -57,13 +75,22 @@ public String getFullname() {
+	 * @return true if the user authenticated with MFA, false if not, null if MFA status 
+	 * is unknown or not supported by the provider.
+	 */
+	public Boolean getMfaAuthenticated() {
+		return mfaAuthenticated;
+	}
```
I would think hard about whether we need the distinction between null and false here, but if null really is needed (not sure why it would be) then this should return an Optional

MrCreosote replied:
There's other places in the code like this

dauglyon replied:
See other comment, will resolve this and we can discuss there.

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/token/StoredToken.java:
```diff
@@ -239,6 +263,7 @@ private static class Builder implements LifeStep, OptionalsStep {
 		private final UserName userName;
 		private Instant creationDate;
 		private Instant expirationDate;
+		private Boolean mfaAuthenticated;
```
I would explicitly set this to the default to make things clear

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -228,6 +233,58 @@ private OrcIDAccessTokenResponse(
+				try {
+					// JWT format: header.payload.signature
+					final String[] parts = idToken.split("\\.");
+					if (parts.length != 3) {
+						// Invalid JWT format
+						return null;
+					}
```
We should assume that OrcID isn't handing out malformed tokens and throw a RuntimeError if this happens.

MrCreosote replied:
Same for the other cases below

MrCreosote replied:
I think this line in the docs is outdated:

> - Malformed JWT during login: logs warning and defaults to MfaStatus.UNKNOWN for graceful degradation

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -228,6 +233,58 @@ private OrcIDAccessTokenResponse(
+				
+				try {
+					// JWT format: header.payload.signature
+					final String[] parts = idToken.split("\\.");
```
is this copied from the Google provider? I seem to remember something like this. I'm pretty sure I wasn't happy about how that was implemented; it'd be better to use a library if possible

dauglyon replied:
Modified from some Orcid openid examples; will look for a library

MrCreosote replied:
I would look, but timebox yourself to an hour or two is my advice. I remember looking but not finding one that was reasonably simple to use

dauglyon replied:
Seems most libraries that do this are pretty heavy (eg, auth0) and focus on creation and validation of tokens. Looking online, manual parsing seems to be common in Java for just reading the claims. But, [auth0/java-jwt](https://github.com/auth0/java-jwt) is an option. I think the hand-rolled code is probably fine here, unless we want to start verifying the signature, or something.

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -228,6 +233,58 @@ private OrcIDAccessTokenResponse(
 				this.accessToken = accessToken.trim();
 				this.fullName = fullName == null ? null : fullName.trim();
 				this.orcID = orcID.trim();
+				this.idToken = idToken == null ? null : idToken.trim();
+				this.mfaAuthenticated = parseAmrClaim(this.idToken);
```
This class is a value class and shouldn't do anything else. As such, I would move the parsing method outside of this class and just pass in the mfa boolean. Also idToken is never used so it seems that could be deleted.

dauglyon replied:
This may be Java class lingo I don't understand; what's the bad pattern there? "This" access?

MrCreosote replied:
Doing anything other than just saving the constructor arguments to instance variables. Think of it like a python named tuple

MrCreosote replied:
The logic looks good, but it still has an unused (AFAICT) jwt field

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -228,6 +233,58 @@ private OrcIDAccessTokenResponse(
+			 * to determine if multi-factor authentication was used.
+			 * 
+			 * @param idToken the JWT ID token from ORCID
+			 * @return true if MFA was used, false if password only, null if unknown or parsing failed
```
We should definitely not conflate a provider not supporting mfa and a mfa processing error

dauglyon replied:
From the userspace view it would be equivalent to us not supporting their provider, but I take your point

MrCreosote replied:
Now processing errors all result in an exception

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -228,6 +233,58 @@ private OrcIDAccessTokenResponse(
+			 */
+			private Boolean parseAmrClaim(final String idToken) {
+				if (idToken == null || idToken.trim().isEmpty()) {
+					return null;
```
If the token is expected this should throw an error

dauglyon replied:
This was throwing in tests, but I think fixed those tests, so I bet I can remove this check now

dauglyon replied:
I'm going to keep this check so we don't get runtime errors if ORCID fails to return a JWT, but I am going to log that as happening. The user will see their MFA status as unknown, but the flow won't fail. If the JWT is malformed, however, I am going to throw.

MrCreosote replied:
Unless there's a documented reason oid wouldn't return a token when we've asked for one, this should fail IMO. Don't code for circumstances that can't happen, just fail

dauglyon replied:
The circumstance would be "they change our oAuth scope access" (eg decide to remove our member API access) which can happen without any change on our end. Trying to be a bit defensive. Seems we shouldn't prevent all orcid logins in that case. But if we're ok with that, I can throw here too

ALSO if we throw here, all auth2 deploys will REQUIRE orcid member api access

MrCreosote replied:
Ok. How about a 4th enum value for this case to distinguish it from the provider doesn't support it at all?

MrCreosote replied:
Actually - does changing the scope mean that other auth installs will fail if they don't have member api access?

dauglyon replied:
I'm going to actually have to double check the api on this; it looks like the docs say conflicting things (you need member api to see the jwt claims, but that the openid scope is available on the public api and it's only purpose is to return the jwt) or it's possible it will return two completely different jwts depending on scope or api access... anyway I will check

dauglyon replied:
openid spec says this should throw an error so, thats what we do!

MrCreosote replied:
Currently it's returning UNKNOWN, should it be changed to throw an error?

MrCreosote replied:
this comment is still open I think

dauglyon replied:
Resolved. https://github.com/kbase/auth2/pull/471/files#diff-3c64efab16dd438379e5d34a8903782c03a9685709c008990c59514bfc86b6deR236

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -228,6 +233,58 @@ private OrcIDAccessTokenResponse(
+					}
+					
+					// AMR claim present but in unexpected format
+					return null;
```
Error

MrCreosote replied:
All parse errors now throw

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/identity/RemoteIdentityDetails.java:
```diff
@@ -20,6 +21,22 @@ public RemoteIdentityDetails(
+			final String username,
+			final String fullname,
+			final String email,
+			final Boolean mfaAuthenticated) {
```
This is getting close to needing a builder

MrCreosote replied:
Nothing to do here :+1:

MrCreosote commented on src/main/java/us/kbase/auth2/service/api/Token.java:
```diff
@@ -55,7 +56,8 @@ public Token(final Authentication auth, final UserAgentParser userAgentParser) {
 	@Produces(MediaType.APPLICATION_JSON)
 	public APIToken viewToken(@HeaderParam(APIConstants.HEADER_TOKEN) final String token)
 			throws NoTokenProvidedException, InvalidTokenException, AuthStorageException {
-		final StoredToken ht = auth.getToken(getToken(token));
+		final IncomingToken it = getToken(token);
+		final StoredToken ht = auth.getToken(it);
```
Why this change?

dauglyon replied:
Was for (type) clarity for me, but can change it back

MrCreosote replied:
I don't have a super strong opinion

MrCreosote replied:
The compiler should inline it anyway

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/service/api/APIToken.java:
```diff
@@ -8,21 +8,36 @@ public class APIToken extends ExternalToken {
 	//TODO JAVADOC or swagger
 	
 	private final long cachefor;
+	private final Boolean mfaAuthenticated;
```
This doesn't fit with the other fields in the api. They're all lower case and pretty brief. I'd go with just `mfa`

MrCreosote replied:
:+1:

MrCreosote reviewed (commented):
I haven't looked at tests yet as this review is already pretty huge

EDIT: All comments for this review have been addressed

dauglyon committed:
Change mfa value to enum, from boolean

dauglyon committed:
rename all uses MfaAuthenticated -> mfa

dauglyon committed:
fix tests

dauglyon committed:
rename all uses MfaAuthenticated -> mfa

dauglyon committed:
rename jwt field for clarity

dauglyon committed:
fix tests, equalities

dauglyon committed:
throw errors for jwt parsing issues

dauglyon committed:
add check for null/missing AMR claim, fix tests

dauglyon committed:
Fix hashCode implementations

dauglyon committed:
Fix hascode expectations

dauglyon committed:
Fix test ordering expectation for identity list

dauglyon commented:
See I need another test or two to have complete coverage, will add.

dauglyon committed:
include both scopes, minor refactor

dauglyon committed:
Fix null fullname handling in ORCID OAuth flow

dauglyon committed:
Fix null MFA field handling for backward compatibility

- Add getMfaStatus helper method to handle null MFA fields in database
- Update token and identity deserialization to use helper method
- Add tests for backward compatibility with existing database records
- Prevents NullPointerException when MFA field is missing from stored data

dauglyon committed:
Update ORCID provider tests for OpenID Connect requirements

- Update scope in login URL tests from 'openid' to 'openid+%2Fauthenticate'
- Add JWT tokens to tests that were missing them after OpenID Connect changes
- Update getIdentityWithNoJWT test to expect new error when JWT is missing
- Fix error condition tests to provide valid JWTs so they test intended errors

dauglyon committed:
Add MFA status to all token responses

- Add MFA field to ExternalToken base class with UNKNOWN default
- Remove duplicate MFA handling from APIToken (now inherited)
- Ensures all token endpoints return MFA status for UI compatibility
- Non-ORCID providers default to MFA status UNKNOWN

dauglyon committed:
Replace hashcode assertions with equaility assertions, backed by equalsverifier contract test. this resolves the hardcoded hashcodes causing stochastic test failures depending on execution order/enviroment, test cleanup to follow

dauglyon committed:
Improve ORCID JWT error handling and test robustness

MrCreosote commented on RELEASE_NOTES.md:
According to codecov there's still some uncovered code, which should be coverable since the orcid interactions are mocked and everything else can be unit tested

MrCreosote replied:
This comment still appears to be relevant - there's an uncovered error condition and some partials

https://app.codecov.io/gh/kbase/auth2/pull/471

MrCreosote replied:
There's still one partial that's uncovered: https://app.codecov.io/gh/kbase/auth2/pull/471#5d09c22eefde74996144f51bde86072a-R237

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
```diff
@@ -745,13 +746,19 @@ public void forceResetAllPasswords(final IncomingToken token)
+	}
+	
+	private NewToken login(final UserName userName, final TokenCreationContext tokenCtx, 
+			final MfaStatus mfa) throws AuthStorageException {
```
```suggestion
	private NewToken login(
			final UserName userName,
			final TokenCreationContext tokenCtx, 
			final MfaStatus mfa
			) throws AuthStorageException {
```

My rule is either all args on 1 line or each arg on its own line. More readable

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/identity/MfaStatus.java:
You might want to rework this in a similar way to TokenType so that we can change the enum name without breaking things later:

https://github.com/kbase/auth2/blob/47f6a0160a6c3bec09f49a6623b3decb3d647a8d/src/main/java/us/kbase/auth2/lib/token/TokenType.java

dauglyon replied:
I *think* i did what you meant here.

MrCreosote replied:
to match the other enums the name should be all caps and the ID SnakeCase

MrCreosote replied:
Also the fromID method accepting anything seems dangerous - easy to introduce a bug. It should only accept valid ID strings unless there's a really good reason not to do that

dauglyon replied:
I can make the ID snakecase but the existing tokens saved in CI are all-caps. I propose I make fromID accept both casings, just in case.

MrCreosote replied:
nah, we can just have people log in again. Let's not make a special case for like 3 people that will resolve itself on login

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/identity/RemoteIdentityDetails.java:
```diff
@@ -36,6 +51,7 @@ public RemoteIdentityDetails(
 		} else {
 			this.email = email.trim();
 		}
+		this.mfa = mfa;
```
```suggestion
		this.mfa = requireNonNull(mfa, "mfa");
```

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/identity/RemoteIdentityDetails.java:
```diff
@@ -57,13 +73,21 @@ public String getFullname() {
 		int result = 1;
 		result = prime * result + ((email == null) ? 0 : email.hashCode());
 		result = prime * result + ((fullname == null) ? 0 : fullname.hashCode());
+		result = prime * result + ((mfa == null) ? 0 : mfa.name().hashCode());
```
```suggestion
		result = prime * result + ((mfa == null) ? 0 : mfa.hashCode());
```

Getting the hash directly from the enum class is fine, it's done this way in other places in the codebase.

There's a couple other places in the PR where this should be fixed

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/storage/mongo/MongoStorage.java:
```diff
@@ -568,6 +569,10 @@ private DisplayName getDisplayName(final String displayName) throws AuthStorageE
 	}
 	
+	private MfaStatus getMfaStatus(final String mfaString) {
+		return mfaString != null ? MfaStatus.valueOf(mfaString) : MfaStatus.UNKNOWN;
```
```suggestion
	private MfaStatus getMfaStatus(final String mfaString) {
		// for backwards compatibility with versions of auth older than 0.7.2
		return mfaString != null ? MfaStatus.valueOf(mfaString) : MfaStatus.UNKNOWN;
```

MrCreosote replied:
If you haven't already, can you add tests for the case where there are remoteidentity and token documents without mfa fields? Should just be unit tests on this class

MrCreosote replied:
I think this one is still relevant

dauglyon replied:
Tests now exist for this

dauglyon replied:
Do you think the comment is needed?

dauglyon replied:
If so feel free to commit the suggestion, should cause zero changes functionally. backwards compatibility is tested

MrCreosote replied:
yeah,  someone in the future may see that mfaString can't be null in the current code and try to remove the ternary operator. Unfortunately I can't commit the comment

dauglyon replied:
resolved.

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/storage/mongo/MongoStorage.java:
```diff
@@ -1596,7 +1603,8 @@ private void updateIdentity(final RemoteIdentity remoteID)
 				.append(pre + Fields.IDENTITIES_EMAIL, rid.getEmail())
-				.append(pre + Fields.IDENTITIES_NAME, rid.getFullname()));
+				.append(pre + Fields.IDENTITIES_NAME, rid.getFullname())
+				.append(pre + Fields.IDENTITIES_MFA, rid.getMfa().name()));
```
This is updating the user document, which I don't think you want to do

dauglyon replied:
True, but I'm going to leave comments why just this field is excluded

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/token/StoredToken.java:
```diff
@@ -269,10 +295,16 @@ public OptionalsStep withContext(final TokenCreationContext context) {
 
+		@Override
+		public OptionalsStep withMfa(final MfaStatus mfa) {
+			this.mfa = mfa;
```
```suggestion
			this.mfa = requireNonNull(mfa, "mfa");
```

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
```diff
@@ -3142,6 +3151,7 @@ public long getSuggestedTokenCacheTime() throws AuthStorageException {
 		return cfg.getAppConfig().getTokenLifetimeMS(TokenLifetimeType.EXT_CACHE);
 	}
 	
+	
 	/** Get the external configuration without providing any credentials.
```
```suggestion
	}
	
	/** Get the external configuration without providing any credentials.
```

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/service/common/ExternalToken.java:
```diff
@@ -28,6 +30,8 @@ public ExternalToken(final StoredToken storedToken) {
 		created = storedToken.getCreationDate().toEpochMilli();
 		custom = storedToken.getContext().getCustomContext();
+		// For tokens from non-ORCID providers, MFA status defaults to UNKNOWN
+		mfa = storedToken.getMfa() != null ? storedToken.getMfa() : MfaStatus.UNKNOWN;
```
```suggestion
		mfa = storedToken.getMfa()
```

Can't be null with the other changes in this review

MrCreosote replied:
:+1:

MrCreosote commented on src/test/java/us/kbase/test/auth2/lib/AuthenticationTokenTest.java:
```diff
@@ -1187,4 +1187,5 @@ public void deleteLoginOrLinkStateNull() throws Exception {
 		}
 	}
 	
+	
 }
```
```suggestion
	}
	
}
```

MrCreosote replied:
:+1:

MrCreosote reviewed (commented):
Review is already huge so skipping tests until next time

EDIT: All comments addressed

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -68,6 +79,7 @@ public static class OrcIDIdentityProvider implements IdentityProvider {
 		private static final Client CLI = ClientBuilder.newClient();
 		
 		private static final ObjectMapper MAPPER = new ObjectMapper();
+		private static final org.slf4j.Logger LOGGER = LoggerFactory.getLogger(OrcIDIdentityProviderFactory.class);
```
Can you handle logging like this:

https://github.com/kbase/auth2/blob/47f6a0160a6c3bec09f49a6623b3decb3d647a8d/src/main/java/us/kbase/auth2/lib/Authentication.java#L325-L327

There's some weirdness I've encountered in the past where a static logger could cause (I think) permgen OOMs. Probably won't happen here but just to be safe

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -256,10 +332,15 @@ private OrcIDAccessTokenResponse getAccessToken(
 					(String) m.get("access_token"),
-					(String) m.get("name"),
-					(String) m.get("orcid"));
+					fullName != null ? fullName : "",
```
Why this change? It's backwards incompatible and all the other providers return null if there's no full name

dauglyon replied:
A mistake, artifact from when I was trying to use only one auth call

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -205,17 +218,78 @@ public void handleError(final Response r, final Map<String, Object> response)
+		 * to determine if multi-factor authentication was used.
+		 *
+		 * @param jwt the JWT ID token from ORCID (may be null/empty for non-member accounts)
+		 * @return MfaStatus indicating whether MFA was used, UNKNOWN if JWT is missing or unparseable
```
```suggestion
		 * @return MfaStatus indicating whether MFA was used, UNKNOWN if JWT is missing
```

although from a comment from review # 1, it sounds like missing should throw an error?

dauglyon replied:
I couldn't figure out how to make the error informative to the user and not just crap-out without a lot of changes to other parts of the code. Perhaps an mfa:ERROR state? Right now I'm just printing to logs that the error occurred

MrCreosote replied:
So I'm still a little hazy - is this never supposed to happen under any circumstances? If so, just crap out

dauglyon replied:
Ok, I solved this with [aa2a416](https://github.com/kbase/auth2/pull/471/commits/aa2a41676d6ff17c24ee0be07f67eb97c174a37b)

This way, we can throw an error in this situation, but still deploy with non-member orcid credentials (if we set a config variable)

MrCreosote replied:
So if I understand correctly:

MFA not configured for provider -> this method is skipped
MFA configured but bad JWT -> exception
MFA configured, good JWT -> NOT_USED or USED

Right?

dauglyon replied:
In the commit I referenced above, I added a new config variable "orcid-mfa-enabled" (default true). 

So now, this is the behavior 

MFA not configured for provider OR o-m-enabled=false -> skips checks, MFA.UNKNOWN 
MFA configured but bad/missing JWT -> exception
MFA configured, good JWT -> NOT_USED or USED

This should resolve the ambiguity I was introducing by wanting to support missing JWTs in case we redeploy-without/loose member api access (where we wouldn't get the jwt). Now it errors for all unexpected JWT states, unless we specifically disable the MFA check, in which case MFA will always be UNKNOWN and the JWT won't be parsed.

dauglyon replied:
I will check that I've updated the rest of the code comments before considering this thread done, because I missed at least one.

MrCreosote replied:
> MFA not configured for provider OR o-m-enabled=false

I'm not sure I understand the difference here - are you talking about the non-OrcID providers re "not configured"

dauglyon replied:
Yeah. non-orcid OR Orcid with the config variable set to false will have mfa UNKNOWN. if the config variable is set to true and orcid is used, the only possible MFA states are USED or NOT_USED (or an error). So, UKNOWN no longer represents any error states

MrCreosote replied:
Ok sounds good. Leaving this unresolved for now per your prior commend

MrCreosote replied:
I'm taking the rocketship to mean that you checked everything and it's all updated at this point

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -205,17 +218,78 @@ public void handleError(final Response r, final Map<String, Object> response)
+					@SuppressWarnings("unchecked")
+					final List<String> amrList = (List<String>) amrClaim;
+					return amrList.contains("mfa") ? MfaStatus.USED : MfaStatus.NOT_USED;
+				} else if (amrClaim instanceof String) {
+					// ORCID may return single string - handle as fallback
+					return "mfa".equals(amrClaim) ? MfaStatus.USED : MfaStatus.NOT_USED;
+				}
+				
+				// AMR claim present but in unexpected format
+				throw new IdentityRetrievalException("AMR claim from ORCID in unexpected format: " + amrClaim);
+				
+			} catch (IllegalArgumentException e) {
+				// Base64 decoding failed - invalid JWT format
+				LOGGER.warn("Unable to decode JWT from ORCID: {}", e.getMessage());
+				throw new IdentityRetrievalException("Unable to decode JWT from ORCID: " + e.getMessage(), e);
+			} catch (IOException e) {
+				// JSON parsing failed - malformed payload
+				LOGGER.warn("Unable to parse JWT payload from ORCID: {}", e.getMessage());
+				throw new IdentityRetrievalException("Unable to parse JWT payload from ORCID: " + e.getMessage(), e);
+			}
```
Can you move these up to specifically catch errors from the lines that throw them? Ensures we don't accidentally mask unexpected errors

MrCreosote replied:
:+1:

MrCreosote reviewed (commented):
Sorry, hit the submit button too early

EDIT: All comments addressed

MrCreosote commented:
It's been a while since the first time I reviewed it but I get the feeling the updated implementation is much more straightforward

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
I realize we're going to need to add a way to set the mfa status for the testmode token endpoint as well

dauglyon replied:
Gave this a shot in [861f89b](https://github.com/kbase/auth2/pull/471/commits/861f89b6880c5f767104d2a5e87dcd88d7e34627)

MrCreosote replied:
:+1:

dauglyon committed:
add missing test

dauglyon committed:
Remove MFA persistence from identity documents; make MFA session-only data

dauglyon committed:
Remove unnecessary fullName backwards compatibility in ORCID provider

dauglyon committed:
Update Authentication.java

Co-authored-by: MrCreosote <MrCreosote@users.noreply.github.com>

dauglyon committed:
Refactor: code quality improvements from PR review

- Fix whitespace formatting (Authentication.java, AuthenticationTokenTest.java)
- Remove unnecessary null handling in ExternalToken (mfa can't be null)
- Reformat login() method arguments for consistency
- Simplify hashCode() methods to use enum.hashCode() directly
- Add backwards compatibility comment in MongoStorage.getMfaStatus()
- Remove unused jwt field from OrcIDAccessTokenResponse
- Replace static logger with inline LoggerFactory calls (avoid permgen issues)
- Update JWT error handling documentation (now throws instead of defaulting)
- Fix identity ordering in getMeMaximalInput test (stochastic test issue)

dauglyon committed:
Refactor MfaStatus enum to follow TokenType/Role pattern

- Change enum values from ALL_CAPS to CamelCase (Used, NotUsed, Unknown)
- Add ID field for database persistence (decouples enum name from DB value)
- Add description field for human-readable output
- Add fromID() static method for backwards-compatible deserialization
- Replace valueOf() calls with fromID() in MongoStorage
- Update all references throughout codebase (16 files)

dauglyon committed:
Narrow exception handling in parseAmrClaim

- Move catch blocks closer to Base64.decode() and MAPPER.readValue()
- Prevent masking validation errors

dauglyon committed:
Add backwards compatibility test for identity documents without MFA field

- Test getUserByRemoteIdWithoutMfaField() explicitly removes MFA field from DB
- Verifies MFA defaults to UNKNOWN for old documents
- Complements existing token backwards compatibility test

dauglyon committed:
Support both member and non-member ORCID OAuth applications

- Add orcid-mfa-enabled config (default: true) to support both ORCID account types
- Member apps (true): use openid scope, parse JWT for MFA status, error on missing JWT
- Non-member apps (false): use basic scope, skip JWT, return MfaStatus.Unknown
- Provide clear error messages guiding users to config option when JWT missing
- Update tests to use config option for non-JWT test scenarios

dauglyon committed:
Add MFA status support to testmode token endpoint

- Added optional 'mfa' parameter to CreateTestToken class
- Backwards compatible constructor defaults mfa to null (Unknown)
- Updated Authentication.testModeCreateToken to accept MfaStatus parameter
- Updated tests to include MfaStatus.Unknown in mock expectations

dauglyon committed:
revert storage ID for compat with CI deploy

dauglyon committed:
Update src/test/java/us/kbase/test/auth2/lib/AuthenticationTokenTest.java

Co-authored-by: MrCreosote <MrCreosote@users.noreply.github.com>

dauglyon committed:
Update src/main/java/us/kbase/auth2/lib/identity/RemoteIdentityDetails.java

Co-authored-by: MrCreosote <MrCreosote@users.noreply.github.com>

dauglyon committed:
Update src/main/java/us/kbase/auth2/lib/token/StoredToken.java

Co-authored-by: MrCreosote <MrCreosote@users.noreply.github.com>

dauglyon committed:
Update src/main/java/us/kbase/auth2/lib/Authentication.java

Co-authored-by: MrCreosote <MrCreosote@users.noreply.github.com>

dauglyon committed:
Fix MFA enum to use consistent ID serialization and enforce non-null values

dauglyon commented on src/main/java/us/kbase/auth2/lib/identity/RemoteIdentityDetails.java:
```diff
@@ -2,6 +2,15 @@
+ * class is session-only and is never persisted to identity documents in the database. The MFA
+ * status is extracted during OAuth callbacks from the identity provider and flows through to
+ * token creation, where it is stored on the token itself. When identities are loaded from the
+ * database, the MFA status will always be {@link MfaStatus#UNKNOWN} since it is not stored.
```
```suggestion
 * database, the MFA status will always be MfaStatus.UNKNOWN since it is not stored.
```

MrCreosote replied:
Why the removal of the link?

dauglyon replied:
Not sure why I did that, probably a find/replace went wrong! Undid that removal

MrCreosote replied:
:+1:

dauglyon committed:
Apply suggestion from @dauglyon

MrCreosote commented:
@dauglyon can you resolve the conflicts? I'll review then

dauglyon commented:
> @dauglyon can you resolve the conflicts? I'll review then

resolved @MrCreosote

dauglyon commented:
> > @dauglyon can you resolve the conflicts? I'll review then
> 
> resolved @MrCreosote

hmmm, nevermind my merge went bad, one moment

dauglyon committed:
Fix compilation errors and test failures from develop merge

- Restore accidentally deleted method body in Authentication.getSuggestedTokenCacheTime()
- Add missing requireNonNull import in RemoteIdentityDetails
- Fix missing closing brace in AuthenticationTokenTest
- Update identity ordering expectation in UserEndpointTest to match new behavior

dauglyon commented:
There are a couple (simple) uncovered Error throws, I will address them after review.

MrCreosote commented:
Why not just cover them?

dauglyon commented:
I will, I just didn't want to keep committing things during your review, and they're easy cases (see codecov), was just flagging it 😅

MrCreosote commented:
Just commit it, I'd rather see 100% code coverage and then reivew

dauglyon committed:
fix coverage

MrCreosote commented:
All the comments around storing the mfa info in RemoteIdentity and how it's different than the other info in there make me think that it's probably the wrong place for it to go. I get why you did that, since that's what all the providers return. I'm thinking that the correct design would be to have a wrapper class so the mfa and RI are separated... but on the other hand that's a lot of work for little benefit.

I'm assuming here that this is purely an internal bit of ugliness - it doesn't affect the mongo user record and doesn't affect any API endpoints right?

If so, maybe just add TODOs along with the comments that ideally the MFA info should be separated out of the RI class at some point. 

Thoughts?

dauglyon commented:
> All the comments around storing the mfa info in RemoteIdentity and how it's different than the other info in there make me think that it's probably the wrong place for it to go. I get why you did that, since that's what all the providers return. I'm thinking that the correct design would be to have a wrapper class so the mfa and RI are separated... but on the other hand that's a lot of work for little benefit.
> 
> 
> 
> I'm assuming here that this is purely an internal bit of ugliness - it doesn't affect the mongo user record and doesn't affect any API endpoints right?
> 
> 
> 
> If so, maybe just add TODOs along with the comments that ideally the MFA info should be separated out of the RI class at some point. 
> 
> 
> 
> Thoughts?

Your understanding is correct. I'm OK with TODOs. It might not be tooo much of a hassle to add a wrapper class but it would likely be a lot more small code changes to review and the code change snowball effect has already affected this PR.

dauglyon committed:
Merge 0.7.3 release notes into 0.7.2 since 0.7.2 not yet released

dauglyon committed:
Fix MfaStatus enum to use consistent ID serialization and enforce non-null values

dauglyon committed:
Add Javadoc link tag for MfaStatus.UNKNOWN reference

dauglyon committed:
Clarify OrcID MFA documentation to specify missing JWT also throws error

dauglyon committed:
Fix MfaStatus enum serialization and test expectations

- Changed RemoteIdentityDetails.toString() to use enum name instead of ID for better debugging
- Updated fromID tests to expect IllegalArgumentException for invalid inputs
- Fixed test expectations to use MFA IDs (Used, NotUsed, Unknown) for JSON serialization
- Fixed test to use ID string when creating test tokens via API

dauglyon committed:
Add tests for getLoginURI with MFA enabled/disabled to achieve full branch coverage

dauglyon committed:
Improve test clarity and coverage

dauglyon committed:
revert mistaken test changes

dauglyon committed:
Add test for invalid MFA status in test mode token creation

dauglyon committed:
Add TODO noting MFA field conflates persistent and transient data in RemoteIdentityDetails

MrCreosote commented:
In hindsight we should've split PR this into multiple parts...

dauglyon committed:
add test for empty string JWT failure

MrCreosote commented on src/main/java/us/kbase/auth2/lib/identity/MfaStatus.java:
```diff
@@ -0,0 +1,65 @@
+package us.kbase.auth2.lib.identity;
+
```
```suggestion
package us.kbase.auth2.lib.identity;
// TODO CODE move to lib/token

```

Looking at this I don't think the mfa stuff should be tangled up with the identity code, it's really relevant only to tokens. That can be done later though, we'll just add a TODO

MrCreosote replied:
:+1:

MrCreosote commented on src/test/java/us/kbase/test/auth2/lib/identity/RemoteIdentityTest.java:
For this and the other 3 tests can you add an assert for the mfa status

MrCreosote replied:
:+1:

MrCreosote commented on src/test/java/us/kbase/test/auth2/lib/identity/RemoteIdentityTest.java:
```diff
@@ -206,4 +203,92 @@ public void compareToNullFails() throws Exception {
+	}
+
+	@Test
+	public void remoteDetailsMfaFailWithNullUser() throws Exception {
```
Should add a test for failing with a null mfa status

MrCreosote replied:
:+1:

MrCreosote commented on src/test/java/us/kbase/test/auth2/lib/identity/RemoteIdentityTest.java:
```diff
@@ -206,4 +203,92 @@ public void compareToNullFails() throws Exception {
+		final RemoteIdentityDetails detsUnknown = new RemoteIdentityDetails("user", "full", "email", MfaStatus.UNKNOWN);
+		assertThat("incorrect mfa status", detsUnknown.getMfa(), is(MfaStatus.UNKNOWN));
+		assertThat("toString should contain mfa status", detsUnknown.toString(), containsString("mfa=UNKNOWN"));
+	}
+
+	@Test
+	public void remoteDetailsMfaFailWithNullUser() throws Exception {
+		try {
+			new RemoteIdentityDetails(null, "full", "email", MfaStatus.USED);
+			fail("created bad details with mfa");
+		} catch (IllegalArgumentException e) {
+			assertThat("incorrect exception msg", e.getMessage(),
+					is("username cannot be null or empty"));
+		}
+	}
+
+	@Test
+	public void mfaStatusGetDescription() throws Exception {
```
```suggestion

    //TODO CODE move the mfa status tests into their own file in lib/token

	@Test
	public void mfaStatusGetDescription() throws Exception {
```


Similar to the above comment I think the mfa stuff should be moved to the token package, but a TODO is fine for now

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/storage/mongo/Fields.java:
```diff
@@ -97,6 +97,8 @@ public class Fields {
 	public static final String IDENTITIES_NAME = "fullname";
 	/** The email address of the identity. */
 	public static final String IDENTITIES_EMAIL = "email";
+	/** Whether the identity was authenticated with multi-factor authentication. */
```
```suggestion
	/** 
	 * Whether the identity was authenticated with multi-factor authentication.
	 * Only used during temporary identity storage during the login process
	 */
```

MrCreosote replied:
:+1: although now I see this field is unused and should be removed

MrCreosote commented on src/main/java/us/kbase/auth2/lib/storage/mongo/MongoStorage.java:
```diff
@@ -1575,7 +1583,15 @@ public Optional<AuthUser> getUser(final RemoteIdentity remoteID) throws AuthStor
+							remoteID.getDetails().getFullname(),
+							remoteID.getDetails().getEmail(),
+							MfaStatus.UNKNOWN));
+			b.withIdentity(updatedIdentity);
```
I don't understand the necessity of this change. If the MFA state is never persisted, it'll always be Unknown for this case right? So this is effectively a noop

dauglyon replied:
Ah, yeah. Good point now that we aren't storing it on remote identity

dauglyon replied:
Actually, I think it needs to be there for the getUserByRemoteIdWithMfa test to work with existing comparison funcs... I am going to add a comment here.

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/storage/mongo/MongoStorage.java:
```diff
@@ -1786,10 +1805,13 @@ private Set<RemoteIdentity> toIdentities(final List<Document> ids) {
 					i.getString(Fields.IDENTITIES_NAME),
-					i.getString(Fields.IDENTITIES_EMAIL));
+					i.getString(Fields.IDENTITIES_EMAIL),
+					MfaStatus.UNKNOWN);
```
Given the 2 constructors, is this change necessary?

MrCreosote replied:
:+1:

MrCreosote commented on src/test/java/us/kbase/test/auth2/lib/storage/mongo/MongoStorageTokensTest.java:
```diff
@@ -56,6 +57,7 @@ TokenType.LOGIN, id, new UserName("bar"))
 					.withCustomContext("k1", "v1")
 					.withCustomContext("k2", "v2")
 					.build())
+			.withMfa(us.kbase.auth2.lib.identity.MfaStatus.NOT_USED)
```
Can you update the other tests to try 1 or 2 other states

MrCreosote replied:
I didn't mean make new tests, I just meant alter a couple of the existing tests to use different states.

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
```diff
@@ -2033,16 +2044,19 @@ public NewToken createUser(
 			final TokenName tokenName,
-			final TokenType tokenType)
+			final TokenType tokenType,
+			final us.kbase.auth2.lib.identity.MfaStatus mfa)
```
```suggestion
			final MfaStatus mfa)
```

MrCreosote replied:
:+1:

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
```diff
@@ -2339,7 +2353,8 @@ public NewToken login(
 			}
 		}
-		return login(u.get().getUserName(), tokenCtx);
+		final MfaStatus mfaStatus = ri.get().getDetails().getMfa();
+		return login(u.get().getUserName(), tokenCtx, mfaStatus);
```
I don't understand how this can ever work - the MFA status is coming from the remote identity stored as part of the temporary session data, but the MFA status isn't persisted to the temporary session data in mongo

MrCreosote replied:
:+1:

MrCreosote reviewed (changes requested):
I'm submitting this review for now because I don't understand how the MFA based flow can ever work - either there's a showstopper bug or I'm very confused

dauglyon committed:
add mfa to TempSessionData

dauglyon committed:
Add tests for MFA field in TemporarySessionData

dauglyon committed:
Update MfaStatus.java

Co-authored-by: MrCreosote <MrCreosote@users.noreply.github.com>

dauglyon committed:
Update RemoteIdentityTest.java

Co-authored-by: MrCreosote <MrCreosote@users.noreply.github.com>

dauglyon committed:
Update Fields.java

Co-authored-by: MrCreosote <MrCreosote@users.noreply.github.com>

dauglyon committed:
Update Authentication.java

Co-authored-by: MrCreosote <MrCreosote@users.noreply.github.com>

dauglyon committed:
address review comments

dauglyon committed:
fix mongo storage to always update identities with MFA.Unknown

MrCreosote commented:
Note to self - all comments are addressed at this point except for https://github.com/kbase/auth2/pull/471#discussion_r2525184258

MrCreosote commented on src/main/java/us/kbase/auth2/lib/storage/mongo/Fields.java:
```diff
@@ -97,6 +97,11 @@ public class Fields {
 	public static final String IDENTITIES_NAME = "fullname";
 	/** The email address of the identity. */
 	public static final String IDENTITIES_EMAIL = "email";
+	/** 
+	 * Whether the identity was authenticated with multi-factor authentication.
+	 * Only used during temporary identity storage during the login process
+	 */
+	public static final String IDENTITIES_MFA = "mfa";
```
```suggestion
```

Unused I think

MrCreosote commented on src/main/java/us/kbase/auth2/lib/storage/mongo/Fields.java:
```diff
@@ -163,7 +170,9 @@ public class Fields {
 	/** The type of the error associated with the temporary token. */
 	public static final String TEMP_SESSION_ERROR_TYPE = "errtype";
-	
+	/** The multi-factor authentication status associated with the temporary token. */
```
```suggestion
	/** The multi-factor authentication status associated with the temporary session data. */
```

MrCreosote commented on src/main/java/us/kbase/auth2/lib/identity/MfaStatus.java:
The admintoken, adminusertokens, tokencreate, and tokens mustache templates need to be updated to show MFA status, which probably means that their underlying APIs need to be updated as well

I was pulled off auth before I could add tests for the admin endpoints / templates so they can be tested manually

MrCreosote replied:
Maybe we should do this in another pr...

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
```diff
@@ -745,13 +746,22 @@ public void forceResetAllPasswords(final IncomingToken token)
 	
 	private NewToken login(final UserName userName, final TokenCreationContext tokenCtx)
 			throws AuthStorageException {
+		return login(userName, tokenCtx, MfaStatus.UNKNOWN);
+	}
```
```suggestion
```

Let's delete this method and make it explicit in the calling methods. Looking at the code, I think the createUser method needs to be updated to pass mfa status as well

MrCreosote commented on src/main/java/us/kbase/auth2/lib/Authentication.java:
There don't appear to be unit tests for any of the changes in this file

MrCreosote commented on src/test/java/us/kbase/test/auth2/lib/AuthenticationTestModeTokenTest.java:
```diff
@@ -71,7 +72,7 @@ public void createTokenWithoutName() throws Exception {
 		when(clock.instant()).thenReturn(Instant.ofEpochMilli(10000));
 		when(rand.getToken()).thenReturn("whee");
 		
-		final NewToken nt = auth.testModeCreateToken(new UserName("foo"), null, TokenType.AGENT);
```
Need to add tests for other mfa types. This could be done by just modifying existing tests I think

MrCreosote commented on src/main/java/us/kbase/auth2/providers/OrcIDIdentityProviderFactory.java:
```diff
@@ -47,28 +51,38 @@ public IdentityProvider configure(final IdentityProviderConfig cfg) {
+	 *
+	 * Multi-Factor Authentication (MFA) Status Handling:
+	 * - Uses OpenID Connect JWT tokens to determine MFA status via AMR claims
+	 * - Configuration option "orcid-mfa-enabled" (default: true):
```
This doesn't appear to be documented anywhere. deploy.cfg.example should do it, see the globus id provider for an example

MrCreosote commented on src/main/java/us/kbase/auth2/lib/identity/MfaStatus.java:
The admintoken, adminuser, tokencreate, and tokens mustache templates need to be updated with MFA status, which means their respective endpoints probably need updates as well. I was pulled off auth before I could finish the admin tests so we can test those manually.

MrCreosote commented on src/test/java/us/kbase/test/auth2/lib/storage/mongo/MongoStorageTokensTest.java:
```diff
@@ -56,11 +58,64 @@ TokenType.LOGIN, id, new UserName("bar"))
 	}
-	
+
+	@Test
```
For consolidation purposes I'm repeating a comment from a previous review here - I think rather than making 2 new tests existing tests could be edited to use different mfa statuses

MrCreosote reviewed (changes requested):
My browser just crashed and made it look like my review was gone - there's something odd about the new review UI preview I think. Anyway, I'm submitting this review just in case. More reviewage coming

MrCreosote commented:
This has been implemented in a different set of PRs
